x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -422,49 +402,56 @@ void getUserTokenFromId(String userTokenId, ActionListener<UserToken> listener)
-                }, e -> {
-                    IOUtils.closeWhileHandlingException(in);
-                    listener.onFailure(e);
-                }));
+            if (version.onOrAfter(Version.V_7_1_0)) {
+                // The token was created in a > 7.1.0 cluster so it contains the tokenId as a String
+                String usedTokenId = in.readString();
+                getUserTokenFromId(usedTokenId, listener);
             } else {
-                IOUtils.closeWhileHandlingException(in);
-                logger.debug("invalid key {} key: {}", passphraseHash, keyCache.cache.keySet());
-                listener.onResponse(null);
+                // The token was created in a < 7.1.0 cluster so we need to decrypt it to get the tokenId
+                if (in.available() < MINIMUM_BASE64_BYTES) {
+                    logger.debug("invalid token, smaller than [{}] bytes", MINIMUM_BASE64_BYTES);
+                    listener.onResponse(null);
+                    return;
+                }
+                final BytesKey decodedSalt = new BytesKey(in.readByteArray());
+                final BytesKey passphraseHash = new BytesKey(in.readByteArray());
+                final byte[] iv = in.readByteArray();
+                final BytesStreamOutput out = new BytesStreamOutput();
+                Streams.copy(in, out);
+                final byte[] encryptedTokenId = BytesReference.toBytes(out.bytes());
+                final KeyAndCache keyAndCache = keyCache.get(passphraseHash);
+                if (keyAndCache != null) {
+                    getKeyAsync(decodedSalt, keyAndCache, ActionListener.wrap(decodeKey -> {
+                        if (decodeKey == null) {
+                            // could happen with a token that is not ours
+                            logger.warn("invalid token");
+                            listener.onResponse(null);
+                            return;
+                        }
+                        try {
+                            final Cipher cipher = getDecryptionCipher(iv, decodeKey, version, decodedSalt);
+                            final String tokenId = decryptTokenId(encryptedTokenId, cipher, version);
+                            getUserTokenFromId(tokenId, listener);
+                        } catch (IOException | GeneralSecurityException e) {
+                            // could happen with a token that is not ours
+                            logger.warn("invalid token", e);
+                            listener.onResponse(null);
+                        }
+                    }, listener::onFailure));
+                } else {
+                    logger.debug(() -> new ParameterizedMessage("invalid key {} key: {}", passphraseHash, keyCache.cache.keySet()));
+                    listener.onResponse(null);
+                }
             }
+        } catch (IOException e) {
+            // could happen with a token that is not ours
+            logger.debug("invalid token");
```

Comment:
if someone does an OAuth2 realm, I think this could get messy in the logs. Maybe:

```suggestion
            if (logger.isDebugEnabled()) {
               logger.debug("built in token service unable to decode token", e);
            } else {
                logger.warn("built in token service unable to decode token");
            }
```

Even then it could still be annoying to see in your logs.

Reply:
I would expect someone implementing a custom oAuth2 realm to disable the in-built TokenService. Also, moving forward, if we make TokenService a more complete oAuth2 server, people might not need a custom oAuth2 realm any more !

Reply:
> I would expect someone implementing a custom oAuth2 realm to disable the in-built TokenService

They might not be able to if they use other realms like SAML

Reply:
Thought about this but it will be too tricky to have two SSO realms working in parallel with lazy session initiation

Reply:
Folks, are we sure we want to increase the log level to warn if we could not parse the token?
Given the way we try authn methods, every APIKey authentication would get one such useless log entry.

I will not make this change, unless anyone _sudo_'s.

Reply:
Yes! You're right, good catch TIL